### PR TITLE
Add UsageTrackerClient to distributor

### PR DESCRIPTION
**Pull request by pracucci**
#### What this PR does

In this PR I'm adding `UsageTrackerClient` to distributor. The distributor doesn't use it yet. Next PR the distributor will use it to enforce the max series limits.

#### Which issue(s) this PR fixes or relates to

N/A

#### Checklist

- [ ] Tests updated.
- [ ] Documentation added.
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`.
- [ ] [`about-versioning.md`](https://github.com/grafana/mimir/blob/main/docs/sources/mimir/configure/about-versioning.md) updated with experimental features.

### CODE REVIEW

File: pkg/usagetracker/tracker.go [hunk 1]
```diff
@@ -21,16 +20,19 @@ import (
...
 	IdleTimeout time.Duration `yaml:"idle_timeout"`
 }
 
 func (c *Config) RegisterFlags(f *flag.FlagSet, logger log.Logger) {
+	f.BoolVar(&c.Enabled, "usage-tracker.enabled", false, "True to enable the usage-tracker.")
+
 	c.InstanceRing.RegisterFlags(f, logger)
 	c.PartitionRing.RegisterFlags(f)
-	f.DurationVar(&c.IdleTimeout, "usage-tracker.idle-timeout", c.IdleTimeout, "The time after which series are considered idle and not active anymore. Must be greater than 0 and less than 1 hour.")
+	f.DurationVar(&c.IdleTimeout, "usage-tracker.idle-timeout", 20*time.Minute, "The time after which series are considered idle and not active anymore. Must be greater than 0 and less than 1 hour.")
```
Comment (pracucci) [inline pkg/usagetracker/tracker.go:35]: @colega I guess this was a typo we introduced in the previous PR. I've set a default of 20m.

File: docs/sources/mimir/configure/configuration-parameters/index.md [hunk 2]
```diff
@@ -1078,6 +1082,88 @@ instance_limits:
...
+  [prefer_availability_zone: <string> | default = ""]
+
+  # (advanced) Delay before initiating requests to further usage-trackers (e.g.
+  # in other zones).
+  # CLI flag: -distributor.usage-tracker-client.requests-hedging-delay
+  [requests_hedging_delay: <duration> | default = 100ms]
+
+  # (advanced) Number of pre-allocated workers used to send requests to
+  # usage-trackers. If 0, no workers pool will be used and a new goroutine will
+  # be spawned for each request.
```
[the remark this answers is not in the document]
Comment (tacole02) [reply docs/sources/mimir/configure/configuration-parameters/index.md:1098]: We avoid using "will" in docs to always maintain present tense.

File: docs/sources/mimir/configure/configuration-parameters/index.md [hunk 3]
```diff
@@ -1078,6 +1082,88 @@ instance_limits:
...
+  # CLI flag: -distributor.usage-tracker-client.reusable-workers
+  [reusable_workers: <int> | default = 500]
+
+  # (advanced) Enable TLS for gRPC client connecting to usage-tracker.
+  # CLI flag: -distributor.usage-tracker-client.tls-enabled
+  [tls_enabled: <boolean> | default = false]
+
+  # (advanced) Path to the client certificate, which will be used for
+  # authenticating with the server. Also requires the key path to be configured.
```
[the remark this answers is not in the document]
Comment (tacole02) [reply docs/sources/mimir/configure/configuration-parameters/index.md:1107]: It's best to use active voice when possible.

File: docs/sources/mimir/configure/configuration-parameters/index.md [hunk 4]
```diff
@@ -1078,6 +1082,88 @@ instance_limits:
...
+  # CLI flag: -distributor.usage-tracker-client.tls-cert-path
+  [tls_cert_path: <string> | default = ""]
+
+  # (advanced) Path to the key for the client certificate. Also requires the
+  # client certificate to be configured.
+  # CLI flag: -distributor.usage-tracker-client.tls-key-path
+  [tls_key_path: <string> | default = ""]
+
+  # (advanced) Path to the CA certificates to validate server certificate
+  # against. If not set, the host's root CA certificates are used.
```
Comment (tacole02) [inline docs/sources/mimir/configure/configuration-parameters/index.md:1117]:
```suggestion
  # against. If not set, it uses the host's root CA certificates.
```

**Comment by pracucci** [comment]
@tacole02 A little suggestion. I think you can skip any PR that's **not** going to be merged to `main`. You can see it here:

Screenshot 2024-12-04 at 09 07 39

This PR is getting merged to a feature branch called `usage-tracker`, which is just used for an Hackathon project. We're **not** going to merge it to `main`.

File: pkg/mimir/modules.go [hunk 5]
```diff
@@ -1176,6 +1223,8 @@ func (t *Mimir) setupModuleManager() error {
 	mm.RegisterModule(TenantFederation, t.initTenantFederation, modules.UserInvisibleModule)
 	mm.RegisterModule(UsageStats, t.initUsageStats, modules.UserInvisibleModule)
 	mm.RegisterModule(UsageTracker, t.initUsageTracker)
+	mm.RegisterModule(UsageTrackerInstanceRing, t.initUsageTrackerInstanceRing)
+	mm.RegisterModule(UsageTrackerPartitionRing, t.initUsageTrackerPartitionRing)
```
Comment (colega) [inline pkg/mimir/modules.go:1227]:
```suggestion
	mm.RegisterModule(UsageTrackerInstanceRing, t.initUsageTrackerInstanceRing, modules.UserInvisibleModule)
	mm.RegisterModule(UsageTrackerPartitionRing, t.initUsageTrackerPartitionRing, modules.UserInvisibleModule)
```

File: pkg/usagetracker/tracker.go (hunk 1 above)
Comment (colega) [reply pkg/usagetracker/tracker.go:35]: Thanks!

File: pkg/mimir/modules.go (hunk 5 above)
Comment (pracucci) [reply pkg/mimir/modules.go:1227]: Oh right, I forgot it!

**Review by colega — APPROVED** [review]

**Pull request merged**